moshi/src/main/java/com/squareup/moshi/StandardJsonAdapters.java

```diff
@@ -215,21 +214,25 @@ static int rangeCheckNextInt(JsonReader reader, String typeMessage, int min, int
 
   static final class EnumJsonAdapter<T extends Enum<T>> extends JsonAdapter<T> {
     private final Class<T> enumType;
-    private final Map<String, T> nameConstantMap = new LinkedHashMap<>();
+    private final Map<String, T> nameConstantMap;
+    private final String[] nameStrings;
 
     public EnumJsonAdapter(Class<T> enumType) {
       this.enumType = enumType;
       try {
         T[] constants = enumType.getEnumConstants();
-        for (T constant : constants) {
+        nameConstantMap = new LinkedHashMap<>();
```

Comment:
The problem with initialCapacity is that it doesn't respect the loadFactor. So if you set it to 100 for a 100 element enum it will still trigger a resize. So you need to account for the default loadFactor when creating initialCapacity and it just gets crazy.